### PR TITLE
[Gradient Compression] Provide a FP16 compression wrapper for DDP communication hooks

### PR DESCRIPTION
**Pull request by iseessel**
Create a FP16 wrapper that can combine FP16 gradient compression with any gradient compression algorithm.

Fixes #{https://github.com/pytorch/pytorch/issues/52986}

### CODE REVIEW

File: test/distributed/test_c10d.py [hunk 1]
```diff
@@ -3870,6 +3895,11 @@ def test_ddp_comm_hook_allreduce_hook_nccl(self):
     def test_default_ddp_comm_hooks_nccl(self):
         self._test_default_ddp_comm_hooks_nccl()
 
+    @requires_nccl()
```
Comment (iseessel) [inline test/distributed/test_c10d.py:3898]: Any advice on how to test my changes? 
Trying to run these locally fails, because I don't have a CUDA-capable GPU.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 2]
```diff
@@ -57,11 +66,36 @@ def fp16_compress_hook(
...
+    """
+
+    def comm_hook_with_fp16(hook_state, bucket: dist.GradBucket):
+        # Cast tensors to FP16. TODO: Optimize this as per https://github.com/pytorch/pytorch/pull/46078/files.
+        bucket_tensors = bucket.get_tensors()
+        bucket_tensors[0] = bucket_tensors[0].to(torch.float16)
+
+        fut = hook(hook_state, bucket)
+
+        # Some hooks are given the ProcessGroup instance directly, whereas others (e.g. powerSGD_hook) are given a state object.
```
Comment (iseessel) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:88]: I think you could consider a refactor by creating a base State class with `processor_group` as an instance variable. The powerSGD hook could then inherit from the base class. It would clean up this code and make it so that all of the hooks accept the same class type as the first arg. 

Although, I thought that would be overkill for this PR.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 3]
```diff
@@ -57,11 +66,36 @@ def fp16_compress_hook(
...
+    """
+    This wrapper applies ``float 32`` to ``float 16`` compression to an arbitrary DDP communication hook.
+
+    Example::
+    >>> state = PowerSGDState(process_group=process_group, matrix_approximation_rank=1, start_powerSGD_iter=10)
+    >>> ddp_model.register_comm_hook(state, fp16_compress_wrapper(powerSGD_hook))
+    """
+
+    def comm_hook_with_fp16(hook_state, bucket: dist.GradBucket):
+        # Cast tensors to FP16. TODO: Optimize this as per https://github.com/pytorch/pytorch/pull/46078/files.
```
Comment (iseessel) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:82]: I think the current implementation would have similar problems as https://github.com/pytorch/pytorch/pull/46078/files. 

I tried to use the `copy_` method here, but didn't have any success, as I think it implicitly converts to float32. See toy example below: 
```
>>> t =  torch.tensor([[1., -1.], [1., -1.]])
>>> t.copy_(t.to(torch.float16))
tensor([[ 1., -1.],
        [ 1., -1.]])
>>> t
tensor([[ 1., -1.],
        [ 1., -1.]])
>>> t.to(torch.float16)
tensor([[ 1., -1.],
        [ 1., -1.]], dtype=torch.float16)
```

Any ideas?

File: test/distributed/test_c10d.py (hunk 1 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3898]: Need to reserve a machine with GPU.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 4]
```diff
@@ -33,6 +33,15 @@ def allreduce_hook(
     """
     return _allreduce_fut(process_group, bucket.get_tensors()[0])
 
+def decompress_fp16_callable(bucket: dist.GradBucket, world_size: int) -> callable:
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:36]: No need to refactor the code and create this util function, because this util function divide the output by `world_size`. Such division should not be used by `fp16_compress_wrapper`, because it's the responsibility of the hook inside the wrapper.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 5]
```diff
@@ -57,11 +66,36 @@ def fp16_compress_hook(
...
-    def decompress(fut):
-        decompressed_tensor = bucket.get_tensors()[0]
-        # Decompress in place to reduce the peak memory.
-        # See: https://github.com/pytorch/pytorch/issues/45968
-        decompressed_tensor.copy_(fut.value()[0].div_(world_size))
-        return [decompressed_tensor]
+    decompress_fp16 = decompress_fp16_callable(bucket, world_size)
+    return fut.then(decompress_fp16)
+
+def fp16_compress_wrapper(hook: callable) -> callable:
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:72]: The output should be `torch.futures.Future`.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 6]
```diff
@@ -57,11 +66,36 @@ def fp16_compress_hook(
...
+
+    def comm_hook_with_fp16(hook_state, bucket: dist.GradBucket):
+        # Cast tensors to FP16. TODO: Optimize this as per https://github.com/pytorch/pytorch/pull/46078/files.
+        bucket_tensors = bucket.get_tensors()
+        bucket_tensors[0] = bucket_tensors[0].to(torch.float16)
+
+        fut = hook(hook_state, bucket)
+
+        # Some hooks are given the ProcessGroup instance directly, whereas others (e.g. powerSGD_hook) are given a state object.
+        try:
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:89]: Since you don't need `world_size` in this wrapper, you don't really need to handle `process_group` here.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 3 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:82]: You are right! In your example, t is created as a FP32 tensor. `copy_` will not change the data type of `t`, so there is an implicit type casting.

You don't need to use `copy_` here, and I think you should be fine with the memory footprint, because the `decompress` part still uses `copy_` to save peak memory.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 5 above)
Comment (iseessel) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:72]: In the example:

```    
>>> state = PowerSGDState(process_group=process_group, matrix_approximation_rank=1, start_powerSGD_iter=10)
    >>> ddp_model.register_comm_hook(state, fp16_compress_wrapper(powerSGD_hook))
```
We have 

```
    def register_comm_hook(self, state: object, hook: callable):
```

so I am pretty sure the wrapper needs to return a callable, whereas the callable needs to return a `torch.futures.Future`. Let me know if I'm missing something!

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 3 above)
Comment (iseessel) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:82]: Do we not get two copies of the tensor here? 

One from the tensors in `bucket_tensors` and one from `bucket_tensors[0].to(torch.float16)` before we overwrite the tensors in `bucket_tensors`?

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 5 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:72]: Yeah, you are right :)

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 3 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:82]: `bucket_tensors[0].to(torch.float16)` is a temporary variable. As long as we use it to overwrite `bucket_tensors` rather than directly return it, i.e., just follow the example of `fp16_compress_hook`, we don't need to worry about memory footprint.

If we directly return `bucket_tensors[0].to(torch.float16)`, then this extra memory will be used for the follow-up process and retained out of the life cycle of this hook.

File: test/distributed/test_c10d.py [hunk 7]
```diff
@@ -3909,6 +3939,11 @@ def test_invalid_powerSGD_state(self):
     def test_default_ddp_comm_hooks_nccl_is_view(self):
         self._test_default_ddp_comm_hooks_nccl(gradient_as_bucket_view=True)
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_default_ddp_comm_hooks_nccl_is_view(self):
```
Comment (SciPioneer) [inline test/distributed/test_c10d.py:3944]: Update the test name

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 3 above)
Comment (iseessel) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:82]: Got it! I just wanted to be extra, extra careful about memory usage, since I know tensors can get pretty big :).

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 8]
```diff
@@ -65,3 +65,31 @@ def decompress(fut):
         return [decompressed_tensor]
 
     return fut.then(decompress)
+
+def fp16_compress_wrapper(hook: callable) -> callable:
+    """
+    This wrapper applies ``float 32`` to ``float 16`` compression to an arbitrary DDP communication hook.
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:71]: Don't mention FP32 here. The input data type can be FP64 or other types, as long as it can be casted into FP16. I will submit another PR to update the docstring of `fp16_compress_hook` as well.

I would reword it like this:
"This wrapper casts the input gradient tensors of a given DDP communication hook to half-precision floating point format (``torch.float16``), and casts the resulting tensors of the given hook back to the input data type such as ``float32``.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 9]
```diff
@@ -65,3 +65,31 @@ def decompress(fut):
...
+
+    Example::
+    >>> state = PowerSGDState(process_group=process_group, matrix_approximation_rank=1, start_powerSGD_iter=10)
+    >>> ddp_model.register_comm_hook(state, fp16_compress_wrapper(powerSGD_hook))
+    """
+
+    def comm_hook_with_fp16(hook_state, bucket: dist.GradBucket) -> torch.futures.Future:
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:78]: Nit: `fp16_compress_wrapper_hook` may be a better name of this hook.

File: test/distributed/test_c10d.py [hunk 10]
```diff
@@ -3815,6 +3815,38 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
             # check whether the grads are equal to what DDP without hook would return.
             self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
 
+
+    def _test_default_ddp_comm_hooks_fp16_wrapper(self, gradient_as_bucket_view=False):
```
Comment (SciPioneer) [inline test/distributed/test_c10d.py:3819]: Please rename it as "_test_fp16_compress_wrapper", and rename the relevant test names accordingly.

File: test/distributed/test_c10d.py [hunk 11]
```diff
@@ -3815,6 +3815,38 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
+        """
+        This unit test verifies whether wrapping the ALLREDUCE and POWER_SGD hooks with
+        the FP16_WRAPPER can give the same result as when there is no hook registered.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        state = process_group
+
+        # test that fp16_compress_hook works with allreduce_hook.
```
Comment (SciPioneer) [inline test/distributed/test_c10d.py:3828]: Nit: `fp16_compress_wrapper` instead of `fp16_compress_hook`

File: test/distributed/test_c10d.py [hunk 12]
```diff
@@ -3815,6 +3815,38 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
+        gpu_model = self._gpu_model_with_ddp_comm_hook(
+            process_group,
+            default.fp16_compress_wrapper(default.allreduce_hook),
+            gradient_as_bucket_view,
+            state
+        )
+        # check whether the grads are equal to what DDP without hook would return.
+        self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
+
+        # test that fp16_compress_hook works with powerSGD comm hook.
```
Comment (SciPioneer) [inline test/distributed/test_c10d.py:3838]: You don't need to write this much boilerplate code. Instead, you can create a loop over both `allreduce_hook` and `powerSGD_hook`.

**Review by wayi1 — COMMENTED** [review]
Looks good overall. Only some minor changes are needed.

File: test/distributed/test_c10d.py (hunk 12 above)
Comment (iseessel) [reply test/distributed/test_c10d.py:3838]: Yeah I thought about that -- but given that they require different states, I thought it would be more clear just to repeat. 

But I don't have any strong feelings either way, so I'll change.

**Comment by iseessel** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/53808.
